spec/java/util/HashSet.main.lsl

```diff
@@ -0,0 +1,633 @@
+            action CALL_METHOD(iter, "hasNext", []),
+            _addAllElements_loop(iter)
+        );
+
+        if (lengthBeforeAdd < this.length)
+        {
+            this.modCount += 1;
+            result = true;
+        }
+        else
+        {
+            result = false;
+        }
+    }
+
+
+    @Phantom proc _addAllElements_loop(iter: Iterator): void
+    {
+        val key: Object = action CALL_METHOD(iter, "next", []);
+        val hasKey: boolean = action MAP_HAS_KEY(this.storage, key);
+
+        if (!hasKey)
+        {
+            action MAP_SET(this.storage, key, HASHSET_VALUE);
+            this.length += 1;
+        }
+    }
+
+
+    @AutoInline proc _throwNPE (): void
+    {
+        action THROW_NEW("java.lang.NullPointerException", []);
+    }
+
+
+    proc _generateKey (visitedKeys: map<Object, Object>): Object
+    {
+        val key: Object = action SYMBOLIC("java.lang.Object");
+        action ASSUME(key != null);
+        val isKeyExist: boolean = action MAP_HAS_KEY(this.storage, key);
+        action ASSUME(isKeyExist);
+
+        val isKeyWasVisited: boolean = action MAP_HAS_KEY(visitedKeys, key);
+        action ASSUME(!isKeyWasVisited);
+        result = key;
+    }
+
+
+    // constructors
+
+    constructor *.HashSet (@target self: HashSet)
+    {
+        this.storage = action MAP_NEW();
+
+        this.length = 0;
+        this.modCount = 0;
+    }
+
+
+    constructor *.HashSet (@target self: HashSet, c: Collection)
+    {
+        this.storage = action MAP_NEW();
+        _addAllElements(c);
+    }
+
+
+    constructor *.HashSet (@target self: HashSet, initialCapacity: int)
+    {
+        if (initialCapacity < 0)
+        {
+            //val initCapStr: String = "Illegal initial capacity: " + action OBJECT_TO_STRING(initialCapacity);
+            action THROW_NEW("java.lang.IllegalArgumentException", []);
+        }
+
+        this.storage = action MAP_NEW();
+
+        this.length = 0;
+        this.modCount = 0;
+    }
+
+
+    constructor *.HashSet (@target self: HashSet, initialCapacity: int, loadFactor: float)
+    {
+        if (initialCapacity < 0)
+        {
+            // val initCapStr: String = "Illegal initial capacity: " + action OBJECT_TO_STRING(initialCapacity);
+            action THROW_NEW("java.lang.IllegalArgumentException", []);
+        }
+
+        //if (loadFactor <= 0 || loadFactor.isNaN) // #problem
+        if (loadFactor <= 0)
+        {
+            // val loadFactorStr: String = "Illegal load factor: " + action OBJECT_TO_STRING(loadFactor);
+            action THROW_NEW("java.lang.IllegalArgumentException", []);
+        }
+
+        this.storage = action MAP_NEW();
+
+        this.length = 0;
+        this.modCount = 0;
+    }
+
+
+    constructor *.HashSet (@target self: HashSet, initialCapacity: int, loadFactor: float, dummy: boolean)
+    {
+        // Problem; We don't have LinkedHashMap automaton at this moment !
+        action TODO();
```

Comment:
@dog-m 
Body of this constructor is such:

```
    HashSet(int initialCapacity, float loadFactor, boolean dummy) {
        map = new LinkedHashMap<>(initialCapacity, loadFactor);
    }
```

What I must write in the body of this constructor ?  @private + action ERROR("Private constructor call"); ?

Reply:
This (our) implementation of `HashSet` does not rely on the `LinkedHashMap` or any other `Map`-interface implementation. Thus, no need to instantiate `LinkedHashMap` here. But its requirements may be checked if needed. Since this is a `private` method it can be called via reflection mechanisms only (I ignore `LinkedHashSet` assuming we will implement it later).
So, marking this constructor with `@private` annotation and using the suggested action (`ERROR`) is a valid solution to my eye.